File: eggrally_v2.ts
// Phase 1 not done gameover logic succ

import { Cmd, h, startModelCmd } from "cs12242-mvu/src"
import { CanvasMsg, canvasView } from "cs12242-mvu/src/canvas"
import * as Canvas from "cs12242-mvu/src/canvas"
import { Match as M, Schema as S, pipe } from "effect"

// TYPES
const Rectangle = S.Struct({
    x: S.Number,
    y: S.Number,
    width: S.Number,
    height: S.Number,
})
type Rectangle = typeof Rectangle.Type

const Egg = S.Struct({
    x: S.Number,
    y: S.Number,
    width: S.Number,
    height: S.Number,
    currentHP: S.Number,
    maxHP: S.Number,
})
type Egg = typeof Egg.Type

const Eggnemy = S.Struct({
    x: S.Number,
    y: S.Number,
    width: S.Number,
    height: S.Number,
    currentHP: S.Number,
    maxHP: S.Number,
})
type Eggnemy = typeof Eggnemy.Type

const Boss = S.Struct({
    x: S.Number,
    y: S.Number,
    width: S.Number,
    height: S.Number,
    currentHP: S.Number,
    maxHP: S.Number,
})
type Boss = typeof Boss.Type

const Entity = S.Union(Egg, Eggnemy, Boss)
type Entity = typeof Entity.Type

const Config = S.Struct({
    screenWidth: S.Int,
    screenHeight: S.Int,
    fps: S.Int,
    canvasID: S.String,
    worldWidth: S.Int,
    worldHeight: S.Int,
    eggWidth: S.Int,
    eggHeight: S.Int,
    eggInitialHP: S.Int,
    eggnemyWidth: S.Int,
    eggnemyHeight: S.Int,
    eggnemyInitialHP: S.Int,
    initialNumberOfEggnemies: S.Int,
    eggnemiesToDefeatForBoss: S.Int,
    bossInitialHP: S.Int,
    bossWidth: S.Int,
    bossHeight: S.Int,
})
type Config = typeof Config.Type

const Model = S.Struct({
    config: Config,
    egg: Egg,
    eggnemies: S.Array(Eggnemy),
    boss: S.Union(Boss, S.Null),
    isGameOver: S.Boolean,
    isWon: S.Boolean,
    score: S.Int,
    ticks: S.Int,
    lastHit: S.Int,
})
type Model = typeof Model.Type

// HELPERS
const getCenter = (e: Entity) => ({
    x: e.x + e.width / 2,
    y: e.y + e.height / 2,
})

const getBounds = (e: Entity) => ({
    top: e.y,
    bottom: e.y + e.height,
    left: e.x,
    right: e.x + e.width,
})

const getSettings = async () => {
    const result = await fetch("settings1.json")
    return result.json()
}

const spawnEggnemies = (settings): Eggnemy[] =>
    Array.from({ length: settings.initialNumberOfEggnemies }, () => {
        const maxX = settings.worldWidth - settings.eggnemyWidth
        const maxY = settings.worldHeight - settings.eggnemyHeight

        return {
            x: Math.floor(Math.random() * maxX),
            y: Math.floor(Math.random() * maxY),
            width: settings.eggnemyWidth,
            height: settings.eggnemyHeight,
            currentHP: settings.eggnemyInitialHP,
            maxHP: settings.eggnemyInitialHP,
        }
    })

const attack = (model: Model): [Eggnemy[], Boss | null, number] => {
    let defeated = 0
    const newEggnemies = model.eggnemies.map((enem) => {
        if (isColliding(model.egg, enem)) {
            const newHP = enem.currentHP - 1
            if (newHP <= 0) defeated++
            return { ...enem, currentHP: newHP }
        }
        return enem
    }).filter(enem => enem.currentHP > 0)

    let newBoss = model.boss
    if (!newBoss && model.score + defeated >= model.config.eggnemiesToDefeatForBoss) {
        newBoss = {
            x: Math.random() * model.config.worldWidth,
            y: Math.random() * model.config.worldHeight,
            width: model.config.bossWidth,
            height: model.config.bossHeight,
            currentHP: model.config.bossInitialHP,
            maxHP: model.config.bossInitialHP,
        }
    } else if (newBoss && isColliding(model.egg, newBoss)) {
        newBoss = { ...newBoss, currentHP: newBoss.currentHP - 1 }
        if (newBoss.currentHP <= 0) newBoss = null
    }

    return [newEggnemies, newBoss, defeated]
}

const move = (egg: Egg, key: string, config: Config) => {
    const speed = 10
    const x = key === "a" ? egg.x - speed : key === "d" ? egg.x + speed : egg.x
    const y = key === "w" ? egg.y - speed : key === "s" ? egg.y + speed : egg.y
    return {
        ...egg,
        x: Math.max(0, Math.min(config.worldWidth - egg.width, x)),
        y: Math.max(0, Math.min(config.worldHeight - egg.height, y)),
    }
}

const moveToward = (from: Entity, to: Entity, speed = 1): Entity => {
    const dx = to.x + to.width / 2 - (from.x + from.width / 2)
    const dy = to.y + to.height / 2 - (from.y + from.height / 2)
    const d = Math.sqrt(dx * dx + dy * dy) || 1
    return {
        ...from,
        x: from.x + (dx / d) * speed,
        y: from.y + (dy / d) * speed,
    } as Entity
}

const isColliding = (a: Entity, b: Entity): boolean => (
    a.x < b.x + b.width && a.x + a.width > b.x &&
    a.y < b.y + b.height && a.y + a.height > b.y
)

const Update = (msg: CanvasMsg, model: Model): Model =>
    M.value(msg).pipe(
        M.tag("Canvas.MsgKeyDown", ({ key }) => {
            if (model.isGameOver || model.isWon) return model
            const egg = move(model.egg, key, model.config)
            const [eggnemies, boss, defeated] = key === "l" ? attack({ ...model, egg }) : [model.eggnemies, model.boss, 0]
            const score = model.score + defeated

            return { ...model, egg, eggnemies, boss, score }
        }),
        M.tag("Canvas.MsgTick", () => {
            let egg = model.egg
            let boss = model.boss
            let lastHit = model.lastHit
            let gameOver = model.isGameOver
            let victory = model.isWon

            const canTakeDamage = (model.ticks + 1 - model.lastHit) >= model.config.fps
            if (canTakeDamage) {
                if (boss && isColliding(egg, boss)) {
                    egg = { ...egg, currentHP: egg.currentHP - 3 }
                    lastHit = model.ticks + 1
                } else {
                    for (const enem of model.eggnemies) {
                        if (isColliding(egg, enem)) {
                            egg = { ...egg, currentHP: egg.currentHP - 1 }
                            lastHit = model.ticks + 1
                            break 
                        }
                    }
                }
            }
            if (egg.currentHP <= 0) gameOver = true
            if (boss && boss.currentHP <= 0) victory = true

            const eggnemies = model.eggnemies.map(enem => moveToward(enem, egg) as Eggnemy)
            if (boss && !victory) boss = moveToward(boss, egg, 2) as Boss

            return {
                ...model,
                egg,
                boss,
                eggnemies: gameOver || victory ? model.eggnemies : eggnemies,
                ticks: model.ticks + 1,
                lastHit,
                isGameOver: gameOver,
                isWon: victory,
            }
        }),
        M.orElse(() => model)
    )

const View = (model: Model) => {
    const offsetX = model.egg.x - model.config.screenWidth / 2
    const offsetY = model.egg.y - model.config.screenHeight / 2
    const time = `${Math.floor(model.ticks / model.config.fps / 60)}:${String(Math.floor(model.ticks / model.config.fps) % 60).padStart(2, '0')}`

    const items = [
        Canvas.Clear.make({ color: "black" }),
        Canvas.SolidRectangle.make({ x: -offsetX, y: -offsetY, width: model.config.worldWidth, height: 1, color: "white" }),
        Canvas.SolidRectangle.make({ x: -offsetX, y: model.config.worldHeight - offsetY, width: model.config.worldWidth, height: 1, color: "white" }),
        Canvas.SolidRectangle.make({ x: -offsetX, y: -offsetY, width: 1, height: model.config.worldHeight, color: "white" }),
        Canvas.SolidRectangle.make({ x: model.config.worldWidth - offsetX, y: -offsetY, width: 1, height: model.config.worldHeight, color: "white" }),
        Canvas.SolidRectangle.make({ 
            x: model.config.screenWidth / 2 - model.egg.width / 2, 
            y: model.config.screenHeight / 2 - model.egg.height / 2,
            width: model.egg.width, 
            height: model.egg.height, 
            color: "white" }),
        Canvas.Text.make({ 
            x: model.config.screenWidth / 2, 
            y: model.config.screenHeight / 2  + model.egg.height / 2 + 12, 
            text: `${model.egg.currentHP}/${model.egg.maxHP}`, 
            fontSize: 12, color: "white", 
            textAlign: "center" }),
        Canvas.Text.make({ x: 50, y: 20, text: `Time: ${time}`, fontSize: 16, color: "white" }),
        Canvas.Text.make({ x: 50, y: 40, text: `Defeated: ${model.score}`, fontSize: 16, color: "white" }),
    ]

    if (model.isWon) {
        items.push(Canvas.Text.make({ x: model.config.screenWidth / 2, y: model.config.screenHeight / 2 - 20, text: "You Win!", fontSize: 20, color: "lime", textAlign: "center" }))
    }

    if (model.isGameOver && !model.isWon) {
        items.push(Canvas.Text.make({ x: model.config.screenWidth / 2, y: model.config.screenHeight / 2 - 20, text: "Game Over", fontSize: 20, color: "red", textAlign: "center" }))
    }

    for (const enem of model.eggnemies) {
    items.push(Canvas.SolidRectangle.make({
        x: enem.x - offsetX,
        y: enem.y - offsetY,
        width: enem.width,
        height: enem.height,
        color: "yellow",
    }))
    items.push(Canvas.Text.make({
        x: enem.x - offsetX + enem.width / 2,
        y: enem.y - offsetY + enem.height + 12,
        text: `${enem.currentHP}/${enem.maxHP}`,
        fontSize: 12,
        color: "yellow",
        textAlign: "center",
    }))
}

    if (model.boss) {
        items.push(Canvas.SolidRectangle.make({ x: model.boss.x - offsetX, y: model.boss.y - offsetY, width: model.boss.width, height: model.boss.height, color: "red" }))
        items.push(Canvas.Text.make({ x: model.boss.x - offsetX, y: model.boss.y - offsetY - 12, text: `${model.boss.currentHP}/${model.boss.maxHP}`, fontSize: 12, color: "red" }))
    }

    return items
}

// START
const root = document.getElementById("app")!

getSettings().then(settings => {
    const initModel: Model = {
        config: { ...settings, canvasID: "canvas" },
        egg: {
            x: (settings.worldWidth / 2) - settings.eggWidth / 2,
            y: (settings.worldHeight / 2) - settings.eggHeight / 2,
            width: settings.eggWidth,
            height: settings.eggHeight,
            currentHP: settings.eggInitialHP,
            maxHP: settings.eggInitialHP,
        },
        eggnemies: spawnEggnemies(settings),
        boss: null,
        isGameOver: false,
        isWon: false,
        score: 0,
        ticks: 0,
        lastHit: -1000,
    }

    startModelCmd(
        root,
        initModel,
        Update,
        canvasView(
            settings.screenWidth,
            settings.screenHeight,
            settings.fps,
            settings.canvasID,
            View
        )
    )
})

File: eggrally_v2.py
import pyxel
import json
import math
import random
from enum import IntEnum
from dataclasses import dataclass
from typing import Callable

# Phase 1 
class PyxelColors(IntEnum):
    BLACK = 0
    DARK_BLUE = 1
    MAGENTA = 2
    CYAN = 3
    BROWN = 4
    BLUE = 5
    INDIGO = 6
    WHITE = 7
    RED = 8
    ORANGE = 9
    YELLOW = 10
    TEAL = 11
    LIGHT_BLUE = 12
    GREY = 13
    PEACH = 14
    PASTEL = 15

@dataclass
class Rectangle:
    x: float
    y: float
    width: int
    height: int
    color: int
    current_hp: int
    max_hp: int

    @property
    def top(self): return self.y
    @property
    def bottom(self): return self.y + self.height
    @property
    def left(self): return self.x
    @property
    def right(self): return self.x + self.width
    @property
    def midx(self): return self.x + self.width // 2
    @property
    def midy(self): return self.y + self.height // 2

@dataclass
class Egg(Rectangle):
    def move(self, speed: int, world_width: int, world_height: int):
        if pyxel.btn(pyxel.KEY_W): self.y -= speed
        if pyxel.btn(pyxel.KEY_S): self.y += speed
        if pyxel.btn(pyxel.KEY_A): self.x -= speed
        if pyxel.btn(pyxel.KEY_D): self.x += speed

        self.x = max(0, min(self.x, world_width - self.width))
        self.y = max(0, min(self.y, world_height - self.height))

@dataclass
class Eggnemy(Rectangle):
    def move_towards(self, egg: Egg, speed: int):
        diffx = egg.midx - self.midx
        diffy = egg.midy - self.midy
        distance = math.hypot(diffx, diffy) or 1
        self.x += diffx / distance * speed
        self.y += diffy / distance * speed

@dataclass
class Boss(Rectangle):
    def move_towards(self, egg: Egg, speed: int):
        diffx = egg.midx - self.midx
        diffy = egg.midy - self.midy
        distance = math.hypot(diffx, diffy) or 1
        self.x += diffx / distance * speed
        self.y += diffy / distance * speed

@dataclass
class Config:
    screen_width: int
    screen_height: int
    world_width: int
    world_height: int
    fps: int
    egg_width: int
    egg_height: int
    egg_init_hp: int
    egg_speed: int
    egg_color: int
    eggnemy_width: int
    eggnemy_height: int
    eggnemy_init_hp: int
    init_eggnemy_count: int
    eggnemies_to_defeat_for_boss: int
    eggnemy_speed: int
    eggnemy_color: int
    boss_width: int
    boss_height: int
    boss_init_hp: int
    boss_speed: int
    boss_color: int

# MVU
class Model:
    def __init__(self, config: Config, egg: Egg, eggnemies: list[Eggnemy]):
        self._config = config
        self._egg = egg
        self._eggnemies = eggnemies
        self._is_game_over = False
        self._score = 0
        self._ticks = 0
        self._last_hit = -1000
        self._boss: Boss | None = None
        self._has_defeated_boss = False

    def update(self):
        if self._is_game_over:
            return

        spawn_chance = 0.2 / self._config.fps
        if random.random() < spawn_chance:
            self._spawn_eggnemy()

        if not self._boss and self._score >= self._config.eggnemies_to_defeat_for_boss:
            self._spawn_boss()

        self._ticks += 1
        self._move_egg()

        can_be_damaged = self._ticks - self._last_hit >= self._config.fps

        for enemy in self._eggnemies:
            enemy.move_towards(self._egg, self._config.eggnemy_speed)
            if self._is_colliding(self._egg, enemy):
                if can_be_damaged:
                    self._egg.current_hp -= 1
                    self._last_hit = self._ticks
                    if self._egg.current_hp <= 0:
                        self._is_game_over = True
        if self._boss:
            self._boss.move_towards(self._egg, self._config.boss_speed)
            if self._is_colliding(self._egg, self._boss):
                if can_be_damaged:
                    self._egg.current_hp -= 3
                    if self._egg.current_hp < 0:
                        self._egg.current_hp = 0
                    self._last_hit = self._ticks
                    if self._egg.current_hp <= 0:
                        self._is_game_over = True

        if pyxel.btn(pyxel.KEY_L):
            self._attack()

    def _move_egg(self):
        offset_x = offset_y = 0
        if pyxel.btn(pyxel.KEY_W): offset_y -= self._config.egg_speed
        if pyxel.btn(pyxel.KEY_S): offset_y += self._config.egg_speed
        if pyxel.btn(pyxel.KEY_A): offset_x -= self._config.egg_speed
        if pyxel.btn(pyxel.KEY_D): offset_x += self._config.egg_speed

        self._egg.x = max(0, min(self._egg.x + offset_x, self.world_width - self._egg.width))
        self._egg.y = max(0, min(self._egg.y + offset_y, self.world_height - self._egg.height))

    def _attack(self):
        attack_range = self._egg.width #?? for relative range?

        new_eggnemies: list[Eggnemy] = []
        for e in self._eggnemies:
            if self._get_distance(self._egg, e) > attack_range:
                new_eggnemies.append(e)
            else:
                assert self._get_distance(self._egg, e) <= attack_range
                e.current_hp -= 1
                if e.current_hp <= 0:
                    self._score += 1
                    continue
                else:
                    new_eggnemies.append(e)

        self._eggnemies = new_eggnemies
        if self._boss:
            if self._get_distance(self._egg, self._boss) <= attack_range:
                self._boss.current_hp -= 1
                if self._boss.current_hp <= 0:
                    self._boss = None
                    self._has_defeated_boss = True
                    self._is_game_over = True

    def _spawn_eggnemy(self):
        self._eggnemies.append(Eggnemy(
            x = random.randint(0, self.world_width - self._config.eggnemy_width),
            y = random.randint(0, self.world_height - self._config.eggnemy_height),
            width = self._config.eggnemy_width,
            height = self._config.eggnemy_height,
            color = self._config.eggnemy_color,
            current_hp = self._config.eggnemy_init_hp,
            max_hp = self._config.eggnemy_init_hp,
        ))
    def _spawn_boss(self):
        self._boss =  Boss(
                x = random.randint(0, self.world_width - self._config.boss_width),
                y = random.randint(0, self.world_height - self._config.boss_height),
                width = self._config.boss_width,
                height = self._config.boss_height,
                color = self._config.boss_color,
                current_hp = self._config.boss_init_hp,
                max_hp = self._config.boss_init_hp,
            )
    
    def _get_distance(self, r1: Rectangle, r2: Rectangle):
        return math.sqrt((r1.midx - r2.midx) ** 2 + (r1.midy - r2.midy) ** 2)

    def _is_colliding(self, r1: Rectangle, r2: Rectangle):
        return not (r1.right < r2.left or r1.left > r2.right or r1.bottom < r2.top or r1.top > r2.bottom)

    @property
    def egg(self): return self._egg
    @property
    def eggnemies(self): return self._eggnemies
    @property
    def boss(self): return self._boss
    @property
    def has_defeated_boss(self): return self._has_defeated_boss
    @property
    def is_game_over(self): return self._is_game_over
    @property
    def score(self): return self._score
    @property
    def fps(self): return self._config.fps
    @property
    def ticks(self): return self._ticks
    @property
    def world_width(self): return self._config.world_width
    @property
    def world_height(self): return self._config.world_height
    @property
    def screen_width(self): return self._config.screen_width
    @property
    def screen_height(self): return self._config.screen_height

class View:
    def __init__(self, screen_width: int, screen_height: int):
        self._screen_width = screen_width
        self._screen_height = screen_height

    def start(self, update: Callable[[], None], draw: Callable[[], None], fps: int):
        pyxel.init(self._screen_width, self._screen_height, fps=fps)
        pyxel.run(update, draw)

    def draw_game(self, model: Model):
        pyxel.cls(0)

        cam_x = model.egg.midx - model.screen_width // 2
        cam_y = model.egg.midy - model.screen_height // 2

        def draw_rect(r: Rectangle):
            pyxel.rect(r.x - cam_x, r.y - cam_y, r.width, r.height, r.color)

        draw_rect(model.egg)
        pyxel.text(model.screen_width // 2 - 10, model.egg.y - cam_y + model.egg.height + 5, f"{model.egg.current_hp}/{model.egg.max_hp}", model.egg.color)

        for e in model.eggnemies:
            draw_rect(e)
            pyxel.text(e.x - cam_x, e.y - cam_y + e.height + 5, f"{e.current_hp}/{e.max_hp}", e.color)

        if model.boss:
            draw_rect(model.boss)
            pyxel.text(model.boss.x - cam_x, model.boss.y - cam_y + model.boss.height + 5, f"{model.boss.current_hp}/{model.boss.max_hp}", PyxelColors.WHITE)

        if model.is_game_over:
            if model.has_defeated_boss and model.egg.current_hp >= 0:
                pyxel.text(model.screen_width // 2 - 15, model.screen_height // 2 - model.egg.height, "You Win!", PyxelColors.WHITE)
            else:
                assert not model.has_defeated_boss and model.egg.current_hp <= 0
                pyxel.text(model.screen_width // 2 - 20, model.screen_height // 2 - model.egg.height, "GAME OVER", PyxelColors.RED)

        pyxel.rectb(-cam_x, -cam_y, model.world_width, model.world_height, PyxelColors.WHITE) # border
        pyxel.text(5, 5, f"Score: {model.score}", PyxelColors.WHITE)
        seconds = model.ticks // model.fps
        pyxel.text(5, 15, f"Time: {seconds//60:02}:{seconds%60:02}", PyxelColors.WHITE)
        
class Controller:
    def __init__(self, model: Model, view: View):
        self._model = model
        self._view = view

    def start(self):
        self._view.start(self.update, self.draw, self._model.fps)

    def update(self):
        self._model.update()

    def draw(self):
        self._view.draw_game(self._model)

if __name__ == "__main__":
    with open("settings1.json") as f:
        s = json.load(f)

    config = Config( # hardcoded speeds and colors for now to work with ts
        screen_width = s["screenWidth"],
        screen_height = s["screenHeight"],
        world_width = s["worldWidth"],
        world_height = s["worldHeight"],
        fps = s["fps"],
        egg_width = s["eggWidth"],
        egg_height = s["eggHeight"],
        egg_init_hp = s["eggInitialHP"],
        egg_speed = 3,
        egg_color = PyxelColors.WHITE,
        eggnemy_width = s["eggnemyWidth"],
        eggnemy_height = s["eggnemyHeight"],
        eggnemy_init_hp = s["eggnemyInitialHP"],
        init_eggnemy_count = s["initialNumberOfEggnemies"],
        eggnemies_to_defeat_for_boss = s["eggnemiesToDefeatForBoss"],
        eggnemy_speed = 1,
        eggnemy_color = PyxelColors.YELLOW,
        boss_width = s["bossWidth"],
        boss_height = s["bossHeight"],
        boss_init_hp = s["bossInitialHP"],
        boss_speed = 2,
        boss_color = PyxelColors.RED,
    )

    egg = Egg(
        x = config.world_width // 2,
        y = config.world_height // 2,
        width = config.egg_width,
        height = config.egg_height,
        current_hp = config.egg_init_hp,
        max_hp = config.egg_init_hp,
        color = config.egg_color
    )

    enemies = [
        Eggnemy(
            x = random.randint(0, config.world_width - config.eggnemy_width),
            y = random.randint(0, config.world_height - config.eggnemy_height),
            width = config.eggnemy_width,
            height = config.eggnemy_height,
            color = config.eggnemy_color,
            current_hp = config.eggnemy_init_hp,
            max_hp = config.eggnemy_init_hp
        )
        for _ in range(config.init_eggnemy_count)
    ]

    model = Model(config, egg, enemies)
    view = View(config.screen_width, config.screen_height)
    controller = Controller(model, view)
    controller.start()
